### ISSUE TITLE
CommandBar doesn't provide overflow for farItems and items cut off at small viewport sizes

### ISSUE THREAD

**Issue by brandonthomas** [microsoft/fluentui#11842]
### Environment Information

- **Package version(s)**: Fabric 7.x.x
- **Browser and OS versions**: NA

### Describe the issue:
Many of our apps use a component we build that includes a CommandBar. For us to be compliant with the reflow part of WCAG 2.1 we need to support viewports down to 320px. The issue with CommandBar and its farItems prop is that it's rendered into an OverflowSet that black boxes a nullRender function. So if we have consumers who put in 10 items it's impossible for them to be compliant. 

Problem Line:
https://github.com/OfficeDev/office-ui-fabric-react/blob/770a4ea294c8d523208bc94532655148d88139f7/packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.base.tsx#L124

Screenshot of cutoff
![image](https://user-images.githubusercontent.com/7529639/73500121-c2b40100-4376-11ea-88f7-7a07be5e203b.png)

### Please provide a reproduction of the issue in a codepen:

https://codepen.io/brandonthomas/pen/KKwjWjm -> Simply go here and squish the viewport to 320px.

#### Actual behavior:
farItems cut off and there's nothing a consumer can do because the null render is black boxed within CommandBar.

#### Expected behavior:
Far items don't cut off when the viewport is reduced.

### Documentation describing expected behavior
https://www.w3.org/TR/WCAG21/#reflow

**Comment by khmakoto** [comment]
Hi @brandonthomas, thanks for filing this issue!

@micahgodbolt and @joschect I think we could change that null render and provide an option for consumers to have overflow items on the right side of the `CommandBar` which would be a backwards compatible change, what do you think?

**Comment by micahgodbolt** [comment]
Yeah, the command bar was built to focus on the primary items. The farside was expected to be so small that it didn't have an overflow behavior. 

To fix this we need to (1) support the far items overflow, and (2) start moving faritems into the overflow once all of the primary items have been moved over.

One thing to consider is that under a given breakpoint, maybe commandbar isn't a good control to use. A bar with 2 overflow menus full of items is....odd.  Maybe under a given width you render the data as a dropdown instead, or some vertical navigation.

**Comment by brandonthomas** [comment]
@micahgodbolt how are we feeling about this? I agree that having a dual overflow is less than ideal but I'm trying to figure out how we address the issue here without putting a ton of burden on all the feature teams. Issue is that even small items in the right side need to overflow when our viewport is only 320px wide.

**Comment by micahgodbolt** [comment]
This needs to be a design driven change. At this point guidance is to use CommandBar at 100% page width and to ensure that far items + overflow aren't too large for the screen. 

If we are to support overflowing far items, we need design to drive those changes.

**Comment by brandonthomas** [comment]
@betrue-final-final is this you then? We need a solution here. Our only other option is to reflow the right items to the left which probably comes with its own baggage.

**Comment by micahgodbolt** [comment]
one thought is that maybe you don't have near/far items at all, and you just have items that are split across the command bar (Outlook asked about this too).

In this case you can always put everything in near items and just push one of the middle items to fill space https://codepen.io/micahgodbolt/pen/KKpeGJp

**Comment by jeffersoneagley** [comment]
+1 on this one, and +1 on the bandaid above.

**Comment by sysofwan** [comment]
+1 this is impacting WCAG 2.1 compliance on our side

**Comment by smhigley** [comment]
One option might be to collapse them both into the same overflow menu, and split that menu into sections with a divider between main items and farItems.

My current questions are:
- which collapses first, main items, or farItems?
- do customers need any control over how farItems are rendered in the overflow menu?

**Comment by brandonthomas** [comment]
My off the cuff thoughts would be far would collapse first by default but maybe have an option to customize this behavior as visual order may not equal priority for the user? In the M365 Admin Design System we use the far items for persistent actions like search or changing to compact/normal mode of the list. Main items are contextually based on selection when used with a list. Going to be hard to guess one way and get it right for everyone IMHO.

In terms of how items are rendered in overflow I think we might. One of our patterns has Search in the command bar and being able to have flexibility there is valuable. I think we'll likely have an option in there that then shows search so it's more a mode with trap focus but I'd need to check.

**Comment by jeffersoneagley** [comment]
Could we have a `disableCollapse` prop on CommandBarItem maybe? I've noticed my CommandBar-based app bar starting to become a catch-all for things we don't have a logical place for a button, but that's another issue entirely 😁

**Comment by brandonthomas** [comment]
Is there any update here @JustSlone @smhigley ?

**Comment by JustSlone** [comment]
It's unlikely that we'll be able to do design update of CommandBar, any design updates would be tackled with the vNext designs. 

Is the bandaid khmakoto mentioned above sufficient for the current needs? 

Then we can consider this issue with a larger refactor/design update of CommandBar.

**Comment by yupasz** [comment]
+1. Resize the page to 400% and change the display resolution to 1280*1024, the search box and the change view button will not be visible
![image](https://user-images.githubusercontent.com/69033452/123930327-391cbe80-d9c2-11eb-881c-bcf123428db2.png)

**Comment by brandonthomas** [comment]
@JustSlone my main concern here is that it's pretty wide spread so then the ask is that every instance has to be patched in this way. On the admin-controls side we can fix our implementation in our composed components but that doesn't really help anyone else. So I think we might still need something as these issues will impact grades for all products.

**Comment by yupasz** [comment]
> ### Environment Information
> * **Package version(s)**: Fabric 7.x.x
> * **Browser and OS versions**: NA
> 
> ### Describe the issue:
> Many of our apps use a component we build that includes a CommandBar. For us to be compliant with the reflow part of WCAG 2.1 we need to support viewports down to 320px. The issue with CommandBar and its farItems prop is that it's rendered into an OverflowSet that black boxes a nullRender function. So if we have consumers who put in 10 items it's impossible for them to be compliant.
> 
> Problem Line:
> https://github.com/OfficeDev/office-ui-fabric-react/blob/770a4ea294c8d523208bc94532655148d88139f7/packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.base.tsx#L124
> 
> Screenshot of cutoff
> ![image](https://user-images.githubusercontent.com/7529639/73500121-c2b40100-4376-11ea-88f7-7a07be5e203b.png)
> 
> ### Please provide a reproduction of the issue in a codepen:
> https://codepen.io/brandonthomas/pen/KKwjWjm -> Simply go here and squish the viewport to 320px.
> 
> #### Actual behavior:
> farItems cut off and there's nothing a consumer can do because the null render is black boxed within CommandBar.
> 
> #### Expected behavior:
> Far items don't cut off when the viewport is reduced.
> 
> ### Documentation describing expected behavior
> https://www.w3.org/TR/WCAG21/#reflow

The reproduction link seems not working.
Updated link: https://codepen.io/yupasz/pen/QWvBgKM

**Comment by JustSlone** [comment]
Maybe I'm blind this afternoon 😄 , but I'm not seeing repro on the updated link?

At 320px (or 1280 @ 400%) I see this which seems reasonable?
![Screen Shot 2021-10-01 at 4 50 42 PM](https://user-images.githubusercontent.com/40304143/135696605-1811e081-12b7-4842-85a6-2fe46aa52ce4.png)
